EventGenerator/src/RPCGun_module.cc

```diff
@@ -56,214 +54,207 @@ namespace mu2e {
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    double weight = 0.;
+    double tau = 0.;
+    const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+    for(const auto& p : *simh) {
+      if(p.second.daughters().empty()) {
+          art::Ptr<SimParticle> part(simh, p.first.asUint());
+          tau = part->endProperTime() / gc.getParticleLifetime(part->pdgId());
+          while(part->parent().isNonnull()) {
+            if((part->creationCode() == ProcessCode::mu2ePrimary)) {
+              part = part->parent();
+              tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId());
             }
-          }else{
-            stop = tstop;
-            break;
+            else {
+              part = part->parent();
+              if ( std::binary_search(decayOffPDGCodes_.begin(), decayOffPDGCodes_.end(), int(part->pdgId()) ) ) {
+                tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId()); 
+              }
           }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
-          }
-        }
-      }else{
-        const auto& tstop = stops_.fire();
-        stop = tstop;
+        } 
+        weight += exp(-tau);
       }
     }
+    if(doHistograms_) _time->Fill(weight);
+    return weight;
+  }
 
-    //std::cout << "Found stop " << exp(-stop.tauNormalized) << " " << stop.t << std::endl;
-
-    const CLHEP::Hep3Vector pos(stop.x, stop.y, stop.z);
-
-    const double energy = generateEnergy();
-
-    if(!generateInternalConversion_){
-      output->emplace_back( PDGCode::gamma,
-                            GenId::ExternalRPC,
-                            pos,
-                            CLHEP::HepLorentzVector( randomUnitSphere_.fire(energy), energy),
-                            stop.t );
-
-      event.put(std::move(output));
-      event.put(std::move(timemap));
-    } else {
+  //================================================================
+  void RPCGun::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    const auto pis = stoppedPiMinusList(simh);
+    double time_weight = MakeEventWeight(event);
+    std::unique_ptr<EventWeight> pw(new EventWeight(time_weight));
+    event.put(std::move(pw)); 
+    if(pis.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"RPCGun::produce(): no suitable stopped pion in the input SimParticleCollection\n";
+    }
+    for(const auto& pistop: pis) {
+      addParticles(output.get(), pistop);
```

Comment:
This code is treating events with multiple stopped pions as producing multiple RPC. I think it should instead pick one stopped pion/event, as the probability of 2 pions stopping and being radiatively captured is 0.  Sampling will complicate the normalization though, as it will underestimate the stopping rate.  Hopefully this is a small correction.

Reply:
So just remove the for loop or use pistop[0]?

Reply:
Or maybe change the return of this https://github.com/Mu2e/Offline/pull/672/files#diff-bb10b320f39c7015a06ded85f65e4b3f674697291c10074d12e5f2fb5cd6e49eR36 to not be a vector at all?

Reply:
The pistops is always length = 1. So we either just make it not a vector or just select the first element here